### ISSUE TITLE
Replacing index.csv with a database

### ISSUE THREAD

**Issue by boramalper** [frrad/skyhub#49]
Would you be interested in replacing index.csv with a database as I believe it would be (1) simpler (2) more reliable (3) faster.

We can consider BoltDB, SQLite or whatever you prefer but an existing & mature solution would be a better idea than our hand-rolled solution I believe.

**Comment by frrad** [comment]
Agreed that CSV is a bit weird. I have experimented with some other index
formats before. You may even be able to find alternative implementations of
the interface still lurking in the code.

I am hoping to eventually allow clients to not have to download the entire
index up front by using a torrent and the same basic trick as in the paper
fetching. It would be nice to use a format which will support that.

I think one of the formats I tried before was boltdb and it had problems
with very bloated file size. I'm afk for the next week or two, will try to
dig up the details for you when I get back.

**Comment by frrad** [comment]
Back at a computer! I agree with the goal of moving away from CSV.

I would like the new format to support
1. Reasonably sized index at rest. Should be no larger than these CSV. Smaller would be better.
1. Loadable via io.Reader so we can use the distribute-as-a-torrent-and-download-as-needed trick.

Regarding BoltDB: 
* Here is the index implementation: https://github.com/frrad/skyhub/blob/master/lib/index/bolt.go#L13
* I can't find the results of my previous experiments, but I recall a pretty big space increase when the db is created. I encourage you to try to reproduce this and see if I was doing something obviously wrong. 

SQLite:
* I think space is unlikely to be an issue, though maybe I'm wrong? Especially as the DB should probably have an index?
* I'm suspicious that it's going to be hard to get this to work on an `io.Reader` instead of a straight file. Again, this is baseless and I would be happy to be proved wrong.

CSV:
* KInd of gross
* Might be able to get the size down by compressing it with something that supports random access (`.xz`?)
* Already can be used through a torrent. The only reason I haven't implemented this is I don't want to have to figure out seeding for the index torrent.

**Comment by boramalper** [comment]
Hi,

Regarding your requirements:
1. I think this is unrealistic. A CSV file does not have an index so queries will be slower, especially if you need random-access, as you suggest in your second point with `io.Reader`.
2. How would this work exactly? Would we publish a new torrent periodically or what?

Regarding the alternatives:

BoltDB:
- My experience with BoltDB wasn't great either... It took a painstakingly long time when I last tried to import.
- Considering that neither of us has any expertise with it, I'd say skip.

SQLite:
- With indices built, it will be larger than the CSV.
- Definitely harder to get work on an `io.Reader`, no doubt. But then again, I don't know how you are planning to do this so I need some more detail. Also, do you think the database is big enough to justify the additional complexity of lazy access?

**Comment by frrad** [comment]
Maybe it would be helpful to step back and explain why I care about small size and lazy access. It's possible that I jumped too quickly to requirements and there are better ways to accomplish my real goal. 

I want users to be able to 
1) Go quickly and easily from installing skyhub to using it. This is why I had in mind to create a torrent containing the index. This will reduce friction around the user having to figure out how to build their own index. The ability to read the index lazily is not a hard requirement for ease-of-use but it would reduce the amount of time it takes for the user's first query to go through. Instead of waiting for the entire index to finish downloading, the user would only have to download log(total # chunks) chunks.
2) Not have skyhub-at-rest consume too much space. Maybe I'm over-indexing on my own case here, but my laptop only has ~25G of space on it and I would like to avoid having to devote too much of that to the skyhub index. I am willing to accept minor slowdowns in return for not taking up huge amounts of space.

    Arguably this is already a lost cause because we're consuming ~60M / article. I consider this to be more acceptable though because the user can always just go in and delete random file chunks and things will continue to work fine / re-download as needed. 

    Being able to download the index lazily would give us this same flexibility for the index. That said, thinking about it more, if the index could be gotten small enough it's possible that the complexity here might outweigh the the advantages. The index compresses very well - I've gotten it to <300M before. If we could get a solution that's compressed at rest and decompresses as needed on the fly it would be hard to argue that the advantages of storing ~5 papers worth of data outweigh the benefits of being able to search the index without the slowdown of going to the network.

#### To your specific objection on how realistic it is to take < size of csv space:

The CSV file is ordered by DOI so it doesn't need an index. Binary search on the CSV file should have the same O(log n) speed as search using a B-Tree database index. There are probably some speed wins from doing less actual reads with a B-Tree but I think they're probably pretty negligible. That said, I don't think it's too unrealistic to be able to get a smaller-than-CSV index. The fact that we're encoding all the IDs as ASCII strings in the CSV seems very inefficient. I think it would be interesting to actually generate the full index in a few formats and see how big it is. Also the data is super compressible.

#### A few more thoughts on  "how would it work exactly" to distribute the the index as a torrent:

Ideally this would work similarly to to the way I have article torrent infohashes now. The skyhub binary would come bundled with the info of some relatively recent index and the user would be able to update indices without re-installing skyhub by editing `torrents.toml`.

It wouldn't be too hard to set up a cron somewhere to regenerate these on a weekly cadence whenever the actual db dumps are updated. I haven't worked on this bit because I feel like there's still a lot of potential for improvement on the index format.

Sorry this turned into a bit of a brain dump... Thoughts?

**Comment by boramalper** [comment]
I just cannot imagine how you can do a binary search on CSV considering that each entry (row) is not of fixed size and thus traversing it would probably require lots of IO. But you have obviously done it already so you are more knowledgable.

I think lazy-access to the index should be considered as a "last" option in case everything else fails as that can increase the latency greatly.

I was thinking of using SQLite with transparent & online compression. There is this blog post (http://blog.ashodnakashian.com/2011/09/sqlite-with-built-in-online-compression/) which demos how it works, and I wonder if we could re-implement it using [lz4](https://github.com/lz4/lz4) for Linux.

Another option is to compress at the filesystem-level using FUSE but that will complicate the installation and distribution of skyhub.

I will keep exploring how to compress the SQLite database in the meanwhile and let you know, but let me know your thoughts and please take a look at [lz4](https://github.com/lz4/lz4), it's blazing fast. =)

**Comment by boramalper** [comment]
I have looked into implementing lz4-compressed SQLite databases and it seems harder than I anticipated, so it's no longer an option I am afraid.

Have you had chance to think about this?

**Comment by frrad** [comment]
Sorry for the slow response. I have been working a bit off and on on a hopefully PR / maybe fork of a library https://github.com/ulikunitz/xz/issues/28#issuecomment-599042430 (some early work on this branch https://github.com/ulikunitz/xz/compare/master...frrad:reader_at?expand=1) which will hopefully allow random access to arbitrary compressed data in the way described in this blog post https://blastedbio.blogspot.com/2013/04/random-access-to-blocked-xz-format-bxzf.html 

This should be easy to use with the current csv thing and hopefully is also generic enough to work with some more interesting file formats also.

**Comment by boramalper** [comment]
There is also a more generic BitTorrent gateway [anacrolix/confluence](https://github.com/anacrolix/confluence/) developed by the same person whose library we are relying on. 

Do you think it might be easier to focus on DOI -> (Torrent, Zip File, File Path) translation and let Confluence handle fetching the file? I will ask if anacrolix would be interested in a zip reader.

**Comment by boramalper** [comment]
You might also find the following interesting: https://github.com/bittorrent/sqltorrent

> Sqltorrent is a custom VFS for sqlite which allows applications to query an sqlite database contained within a torrent. Queries can be processed immediately after the database has been opened, even though the database file is still being downloaded. Pieces of the file which are required to complete a query are prioritized so that queries complete reasonably quickly even if only a small fraction of the whole database has been downloaded.
